src/main/java/edu/umich/ctools/esb/utils/WAPI.java

```diff
@@ -355,24 +356,32 @@ public WAPIResultWrapper reportError(int status) {
 
 	// Wrapper to deal with token renewal and query retry.
 	protected WAPIResultWrapper getOrPutRequestWithRetry(String requestType, String request,HashMap<String,String> headers){
-		WAPIResultWrapper wrappedResult = doGetRequest(request,headers);
-		M_log.info("getOrPutRequestWithRetry() called with request type: " + requestType + " request: "+request+" headers: "+headers);
+		WAPIResultWrapper wrappedResult = null;
+		wrappedResult = choosePutOrGetRequestResult(requestType, request, headers, wrappedResult);
+		M_log.info("getOrPutRequestWithRetry() called with request type: " + requestType + " request: " + request + " headers: " + headers);
 
-		if(wrappedResult.getStatus()==HTTP_UNAUTHORIZED){
+		if (wrappedResult.getStatus() == HTTP_UNAUTHORIZED) {
 			wrappedResult = renewToken();
-			if(wrappedResult.getStatus()==HTTP_SUCCESS){
+			if (wrappedResult.getStatus() == HTTP_SUCCESS) {
 				headers = updateAuthHeader(headers);
-				if (GET.equals(requestType)) {
-					wrappedResult = doGetRequest(request,headers);
-				}
-				if (PUT.equals(requestType)) {
-					wrappedResult = doPutRequest(request,headers);
-				}
+				wrappedResult = choosePutOrGetRequestResult(requestType, request, headers, wrappedResult);
 			}
 		}
 		return wrappedResult;
 	}
 
+	private WAPIResultWrapper choosePutOrGetRequestResult(String requestType, String request, HashMap<String, String> headers, WAPIResultWrapper wrappedResult) {
+		M_log.info("choosePutOrGetRequestResult() called with request type: " + requestType);
+		if (GET.equals(requestType)) {
+			return(wrappedResult = doGetRequest(request, headers));
+		}
+		if (PUT.equals(requestType)) {
+			return(wrappedResult = doPutRequest(request, headers));
+		}
+		M_log.info("WAPI: Unrecognized request type: " +requestType);
+		throw new WAPIException("Unrecognized request type: " +requestType);
```

Comment:
I did not see a catch statement for WAPIException inside this diff.

Reply:
WAPI  won't handle it using catch but throwing an exception application needs to handle it and currently, SPE is  handing it in finally. 
```
try {
			assignmentsFromDW = getSPEGrades(speproperties,priorUpdateTime);

			//// Extract grades from JSON the grades for insertion
			ArrayList<HashMap<String,String>> SPEgradeMaps = null;
			try {
				SPEgradeMaps = convertSPEGradesFromDataWarehouseJSON(assignmentsFromDW);
			} catch (JSONException e) {
				M_log.error("exception in converting grades: "+e);
			}

			// get the time of the newest grade so can store that and only query for newer grades
			// in the future.

			//// Insert the grades.
			M_log.info("Grade count since {} is {}.",priorUpdateTime,SPEgradeMaps.size());
			Instant lastTestTakenTime = putSPEGrades(SPEgradeMaps,priorUpdateTime);
			persisttimestamp.writeLastTestTakenTime(lastTestTakenTime);

		} catch (GradeIOException e1) {
			M_log.error("Exception processing SPE grades:",e1);
		}
		finally {
			// Finish up, save data, send reports.
			closeUpShop();
		}
```